src/Middleware/HttpLogging/src/HttpLoggingMiddleware.cs

```diff
@@ -0,0 +1,289 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Buffers;
+using System.Collections.Generic;
+using System.IO;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.Http.Features;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Options;
+using Microsoft.Net.Http.Headers;
+
+namespace Microsoft.AspNetCore.HttpLogging
+{
+    /// <summary>
+    /// Middleware that logs HTTP requests and HTTP responses.
+    /// </summary>
+    public class HttpLoggingMiddleware
+    {
+        private readonly RequestDelegate _next;
+        private readonly ILogger _logger;
+        private HttpLoggingOptions _options;
+        private const int PipeThreshold = 32 * 1024;
+
+        /// <summary>
+        /// Initializes <see cref="HttpLoggingMiddleware" />.
+        /// </summary>
+        /// <param name="next"></param>
+        /// <param name="options"></param>
+        /// <param name="loggerFactory"></param>
+        public HttpLoggingMiddleware(RequestDelegate next, IOptions<HttpLoggingOptions> options, ILoggerFactory loggerFactory)
+
+        {
+            _next = next ?? throw new ArgumentNullException(nameof(next));
+
+            if (options == null)
+            {
+                throw new ArgumentNullException(nameof(options));
+            }
+
+            if (loggerFactory == null)
+            {
+                throw new ArgumentNullException(nameof(loggerFactory));
+            }
+
+            _options = options.Value;
+            _logger = loggerFactory.CreateLogger<HttpLoggingMiddleware>();
+        }
+
+        /// <summary>
+        /// Invokes the <see cref="HttpLoggingMiddleware" />.
+        /// </summary>
+        /// <param name="context"></param>
+        /// <returns></returns>
+        public async Task Invoke(HttpContext context)
+        {
+            if ((HttpLoggingFields.Request & _options.LoggingFields) != HttpLoggingFields.None)
+            {
+                var list = new List<KeyValuePair<string, object?>>();
+
+                var request = context.Request;
+
+                if (_options.LoggingFields.HasFlag(HttpLoggingFields.Protocol))
```

Comment:
What are the performance characteristics of HasFlag? I recall that one of these methods use to be very bad (IsDefined?) but my knowledge here is very old.

Reply:
HasFlag is efficient now AFAIK but you can check sharplab

Reply:
It is since .NET Core 2.1 an intrinsic and fast.
Cf. [Performance Improvements in .NET Core 2.1](https://devblogs.microsoft.com/dotnet/performance-improvements-in-net-core-2-1/) (search for _HasFlag_ to see the benchmark).